### ISSUE TITLE
Convert Canvas to NAPI 

### ISSUE THREAD

**Issue by aruneshchandra** [nodejs/abi-stable-node#76]

**Comment by aruneshchandra** [comment]
Depends on #50 Typed Arrays

**Comment by jasongin** [comment]
I've already partially ported the canvas package to NAPI as part of #41; see related information there. The remaining work items are:

 1. Update already-ported canvas code for recent NAPI changes (error-handling)
 2. Add typed-arrays support to NAPI (#50)
 3. Port the small parts of canvas code that use typed-arrays to NAPI.
 4. Validate the port using canvas's tests, and fix any bugs caused by porting.

**Comment by jasongin** [comment]
I updated the Canvas port to use the new [NAPI C++ wrapper classes](https://github.com/nodejs/abi-stable-node/pull/88). It still needs typed-array support.

**Comment by jasongin** [comment]
With typed arrays in the NAPI C API (#90) and the wrapper (#88), I was able to complete the canvas port. All the canvas tests are now passing on the NAPI port! I will try benchmarking it next.

**Comment by jasongin** [comment]
Canvas performance testing is tracked as a separate task (#77), so this one can be closed.

The code is at https://github.com/jasongin/node-canvas/tree/napi2

**Comment by jasongin** [comment]
While trying to benchmark canvas, I found a bug that is related to some `Nan::ObjectWrap::Ref()` / `UnRef()` calls that I'd forgotten I had commented out when porting the code. The lack of references led to the object being deleted only when under memory pressure, causing benchmarks to crash.

Unfortunately as I investigated further into this area of persistent references and weak references, I realized that current NAPI APIs related to those concepts are insufficient. I think I've heard others mention that, though I don't see any existing issue tracking it, so I'll open one.

Meanwhile I'm reopening this issue, since a fully-working and benchmarkable canvas port will depend on fixing the NAPI referencing APIs.
